python/ml4ir/applications/ranking/model/ranking_model.py

```diff
@@ -56,15 +118,72 @@ def predict(
+        if logs_dir:
+            outfile = os.path.join(logs_dir, RelevanceModelConstants.MODEL_PREDICTIONS_CSV_FILE)
+            # Delete file if it exists
+            self.file_io.rm_file(outfile)
 
-        return super().predict(
-            test_dataset=test_dataset,
+        _predict_fn = get_predict_fn(
+            model=self.model,
+            tfrecord_type=self.tfrecord_type,
+            feature_config=self.feature_config,
             inference_signature=inference_signature,
+            is_compiled=self.is_compiled,
+            output_name=self.output_name,
+            features_to_return=self.feature_config.get_features_to_log(),
             additional_features=additional_features,
-            logs_dir=logs_dir,
-            logging_frequency=logging_frequency,
+            max_sequence_size=self.max_sequence_size,
         )
 
+        predictions_df_list = list()
+        batch_count = 0
+        # defining variables to compute running mean and variance for t-test computations
+        agg_count, agg_mean, agg_M2 = 0, 0, 0
+        for predictions_dict in test_dataset.map(_predict_fn).take(-1):
+            predictions_df = pd.DataFrame(predictions_dict)
+            if logs_dir:
+                np.set_printoptions(
+                    formatter={'all': lambda x: str(x.decode('utf-8')) if isinstance(x, bytes) else str(x)},
+                    linewidth=sys.maxsize, threshold=sys.maxsize)  # write the full line in the csv not the truncated version.
+                for col in predictions_df.columns:
+                    if isinstance(predictions_df[col].values[0], bytes):
+                        predictions_df[col] = predictions_df[col].str.decode('utf8')
+
+                if os.path.isfile(outfile):
+                    predictions_df.to_csv(outfile, mode="a", header=False, index=False)
+                else:
+                    # If writing first time, write headers to CSV file
+                    predictions_df.to_csv(outfile, mode="w", header=True, index=False)
+
+                # Accumulating statistics for t-test calculation
+                clicked_records = predictions_df[predictions_df[self.feature_config.get_label("node_name")] == 1.0]
+                diff = (clicked_records[RankingConstants.NEW_RANK] - clicked_records[self.feature_config.get_rank("node_name")]).to_list()
+                agg_count, agg_mean, agg_M2 = compute_stats_from_stream(diff, agg_count, agg_mean, agg_M2)
+
+            else:
+                predictions_df_list.append(predictions_df)
+
+            batch_count += 1
+            if batch_count % logging_frequency == 0:
+                self.logger.info("Finished predicting scores for {} batches".format(batch_count))
+
+        predictions_df = None
+        if logs_dir:
+            self.logger.info("Model predictions written to -> {}".format(outfile))
+        else:
+            predictions_df = pd.concat(predictions_df_list)
```

Comment:
@mohazahran Why is this whole thing being duplicated here? Is this the only new line being added?
```agg_count, agg_mean, agg_M2 = compute_stats_from_stream(diff, agg_count, agg_mean, agg_M2)```

Reply:
I think a better way to do this would be to pass all the aggregate stats you need for the t-test as feature functions into `additional_features`. There should be some stateful variables you can use to update existing values IMO. Let me know if you want to sync up about this. It's been a while for me as well.

Reply:
I found an easier alternative for you - move all of this to `evaluate()` which already does this and is a more logical place for doing a t-test anyway.